File: 프로그래머스/2/42578. 의상/의상.cpp
#include <string>
#include <vector>

#include <unordered_map>

using namespace std;

int solution(vector<vector<string>> clothes) {
    
    // 종류 - 개수 map
    unordered_map<string, int> counts;
    for(auto c: clothes)
        counts[c[1]]++;
    
    // 모든 종류의 의상을 함께 입는 경우의 수를 구한다.
    // 단, 해당 종류의 의상을 아예 입지 않는 경우를 고려하여 (의상 개수 + 1)을 곱해준다.
    int result = 1;
    for(auto& p: counts)
        result *= (p.second+1);
    
    int answer = result-1; // 최소 하나의 의상은 입어야하므로 아무것도 입지 않는 경우의 수 하나를 빼준다.
    return answer;
}

File: 프로그래머스/2/42578. 의상/의상.java
import java.util.*;

class Solution {
    public int solution(String[][] clothes) {
        
        Map<String, Integer> mm = new HashMap<>();
        
        // 1. 각 의상 종류별 개수를 센다.
        String key;
        for(String[] cloth: clothes) {
            key = cloth[1];
            if (!mm.containsKey(key))
                mm.put(key, 1);
            else
                mm.put(key, mm.get(key)+1);
        }
        
        // 2. 모든 의상을 조합해 입는 경우의 수를 센다. (각 의상 수를 곱하면 된다.)
        // 단, 해당 의상 종류를 선택하지 않는 경우 하나를 합쳐 곱해준다.
        // ex) 하의 3벌, 상의 2벌을 조합해 입는 경우의 수 -> (3+1) * (2+1)
        int result = 1;
        for(String k: mm.keySet()) {
            result *= (mm.get(k)+1);
        }
        
        // 모든 종류의 의상을 선택하지 않는 경우 하나를 제외한 경우의 수를 리턴한다.
        int answer = result - 1;
        return answer;
    }
}